### DISCUSSION TITLE
Anaglyph shader

### DISCUSSION THREAD

**Discussion by salexx** [StereoKit/StereoKit#1135] [Q&A] [answered]
Hello. Help me understand how to output stereo Anaglyph red cyan image.

I tried everything with
o.view_id = id % sk_view_count
id = id / sk_view_count

in vertex and pixel shader, even just different colors for the left and right eyes do not work)

If possible, a piece of code, or in what direction should I move?

Here are my attempts...

```
psIn vs(vsIn input, uint id : SV_InstanceID)
{
	psIn o;
	o.pos = input.pos;
	o.uv  = input.uv;
	o.cuv = (input.uv-pointer) / cursor_size;// (pointer - input.uv) *cursor_size;

    o.view_id = id % sk_view_count;
    //id = id / sk_view_count;

    id = o.view_id;
    
    o.color = 0;
	
    if (o.view_id == 0)
    {
        o.color.r = 1;
        //anaglyphColor.g = desktop_col.g * 0.7;
        //anaglyphColor.b = 0;
    }

    if (o.view_id != 0)
    {
        //anaglyphColor.r = 0;
        //anaglyphColor.g = 0;
        //anaglyphColor.g = input.col.g;
        o.color.g = 1;
    }

	return o;
}
```

```
float4 ps(psIn input) : SV_TARGET
{
	float4 desktop_col = pow(abs(source.Sample(source_s, input.uv)), 2.2);
	float4 cursor_col  = cursor.Sample(cursor_s, input.cuv);
	float2 bounds      = input.cuv<=1  && input.cuv>=0;
	//return lerp(desktop_col, cursor_col, cursor_col.a*min(bounds.x, bounds.y));
	
    float4 anaglyphColor = 0;

    //uint id = input.view_id % sk_view_count;
    
    //if (id == 0)
    //{
    //    anaglyphColor.r = 1;
    //    anaglyphColor.g = desktop_col.g * 0.7;
    //    anaglyphColor.b = 0;
    //}

    //if (id == 1)
    //{
    //    anaglyphColor.r = 0;
    //    anaglyphColor.g = 0;
    //    anaglyphColor.g = 1;
    //}
	

    //if (id == 2)
    //{
    //    anaglyphColor.r = 0;
    //    anaglyphColor.g = 0;
    //    anaglyphColor.b = 1;
    //}

    //return lerp(anaglyphColor, cursor_col, cursor_col.a * min(bounds.x, bounds.y));
    return input.color;

}
```

**Comment by mvvvv** [comment, marked as answer]
I'm not sure at all but I'd try to substract the opposite color.

**Reply by salexx** [reply]
Good idea. Although I don't know how to do it yet. )

I can't even make changes for the left and right eyes yet.

**Comment by maluoi** [comment]
To _output_ an anaglyph color effect, you actually need 2 renders of a scene that you then combine by taking the red channel from one, and the green/blue channels of the other. This isn't something you can really do just by a single shader! It looks more like a post-processing effect than anything else, and StereoKit doesn't really support post-processing yet.

You can get multiple renders from Renderer.RenderTo, or RenderList.RenderNow and combine those, but displaying it might be a bit troublesome. Maybe use the render layer system to exclude content from the main view, and draw the anaglyph effect onto a fullscreen quad. A bit complicated, but maybe possible.

**Reply by salexx** [reply]
Thanks for the quick reply.
But at least I won't waste time or even try )))
